FlaSpaceInst/EZ-RASSOR
# Add IMU odometry with extended Kalman filter

# Description
This PR includes the following changes:
- Change IMU plugin from `GazeboRosImuSensor` to `GazeboRosImu` because the former did not output covariance values properly (no covariance values were reported for orientation).
- Add an extended Kalman filter for sensor fusion via the `robot_localization` package. The Kalman filter currently only takes in IMU data, but in the future it will read from more sources like wheel odometry and visual odometry.
- Add back the `enable_real_odometry` parameter and use the output of the extended Kalman filter as the odometry source. Since IMU sensors can only reliably estimate orientation, currently when `enable_real_odometry` is set, the output of the Kalman filter is used for the rover's heading, while the rover's actual position in the simulation is used for position. The position measurement will be replaced with the output of the Kalman filter once one of the other odometry sources (which estimate position more reliably) is added.

# Testing
To test these changes, make sure that the rover can still reliably make it to its destination while using the odometry output as the source of its heading estimate. This can be done through the following steps:
- Make sure you have the `robot_localization` package installed by running:
```
sh develop.sh resolve
```
- Run the simulation using odometry by running:
```
roslaunch ezrassor_launcher configurable_simulation.launch control_methods:="autonomy" enable_real_odometry:="true"
```
- In a separate window, run the following to switch to auto-drive mode:
```
rostopic pub /ezrassor1/autonomous_toggles std_msgs/Int8 1
```
- Make sure that the rover eventually reaches its destination (when the rover reaches its destination, it will stop and a message saying "Destination reached!" will be logged).

If you want to see the actual data being published, see the `/ezrassor1/imu` (raw IMU data) and `/ezrassor1/odometry/filtered` (output of the Kalman filter) topics.
